workspace/codespaces/codespaces.py

```diff
@@ -0,0 +1,96 @@
+# with "Codespaces" repository permissions set to "read" and "Organization codespaces"
+# organization permissions set to "read". For more information, see:
+# https://docs.github.com/en/rest/codespaces/organizations?apiVersion=2022-11-28#list-codespaces-for-the-organization
+TOKEN = os.environ["CODESPACES_GITHUB_API_TOKEN"]
+HEADERS = {
+    "Accept": "application/vnd.github+json",
+    "X-GitHub-Api-Version": "2022-11-28",
+}
+
+
+Codespace = collections.namedtuple(
+    "Codespace",
+    ["last_used_days_ago", "owner", "name", "has_unpushed", "has_uncommitted"],
+)
+
+
+def fetch(url, key):
+    """
+    Recursively fetch all records from the GitHub REST API endpoint given by `url`.
+    `key` is the property under which the records are stored.
+    """
+
+    def set_bearer_auth(request):
+        request.headers["Authorization"] = f"Bearer {TOKEN}"
+        return request
+
+    def fetch_one_page(page_url):
+        response = requests.get(page_url, auth=set_bearer_auth, headers=HEADERS)
+        response.raise_for_status()
+        yield from response.json().get(key)
+        if "next" in response.links:
+            yield from fetch_one_page(response.links["next"]["url"])
+
+    yield from fetch_one_page(url)
+
+
+def get_codespace(record):
+    now = datetime.datetime.now(datetime.timezone.utc)
+    last_used_at = datetime.datetime.fromisoformat(record["last_used_at"])
+    last_used_days_ago = (now - last_used_at).days
+    owner = record["owner"]["login"]
+    name = record["name"]
+    has_unpushed = record["git_status"]["has_unpushed_changes"] == "true"
+    has_uncommitted = record["git_status"]["has_uncommitted_changes"] == "true"
+    return Codespace(last_used_days_ago, owner, name, has_unpushed, has_uncommitted)
+
+
+def is_at_risk(codespace, threshold_in_days):
+    return codespace.last_used_days_ago >= threshold_in_days
+
+
+def main():
+    org = "opensafely"
+    threshold_in_days = 20
+
+    records = fetch(URL_PATTERN.format(org=org), "codespaces")
+    codespaces = (get_codespace(rec) for rec in records)
+    at_risk_codespaces = sorted(
+        (cs for cs in codespaces if is_at_risk(cs, threshold_in_days)),
+        key=lambda cs: cs.last_used_days_ago,
+        reverse=True,
+    )
+
+    if at_risk_codespaces:
+        items = [
+            (
+                f"`{cs.owner}` has a Codespace that they last used {cs.last_used_days_ago} days ago "
+                f"({cs.name}).\n"
+                f"Unpushed changes: {'Yes' if cs.has_unpushed else 'No'} | "
+                f"Uncommitted changes: {'Yes' if cs.has_uncommitted else 'No'}\n\n"
+            )
+            for cs in at_risk_codespaces
+        ]
+        body = f"The following `{org}` Codespaces are at risk of deletion.\n\n"
+        body += "".join(items)
+    else:
+        body = f"No `{org}` Codespaces are at risk of deletion :tada:"
+
+    header = "Codespaces Report"
```

Comment:
I'm not sure how to interpret Header/Body here. The output may include many codespaces. If there are things to report can/are they be in a reply to an initial "Codespaces report" message to avoid spamming the channel? (Maybe it is already, I'm just not sure what the bot API is.)

Reply:
I'd suggest either cross-referencing with the screenshots above, or joining BennettTest and setting up a development environment.

Reply:
Thanks, Iain. I asked because I was unclear from the screenshot and didn't know where this was being tested. I've now discovered that BennettTest is a separate Slack workspace available: [here](https://bennetttest.slack.com/) and this was tested in the #random channel there.

Doing so shows that the codespaces notifications are all displayed together in a single message which with 5 workspaces to report takes up most of a single screen in my UI, which is quite a lot of text to push to #team-rex. And it could be longer. So I still think this it would be preferable to have one initial message noting that there are workspaces at risk, and put the message with the details in a reply. Hopefully that's something we can do cost-beneficially with BennettBot's API.